doc/compiler.md

```diff
@@ -0,0 +1,205 @@
+# Using the ProtoDef compiler
+
+The ProtoDef compiler can convert your protocol JSON into javascript code that can read and write buffers directly instead of using the ProtoDef interpreter. Depending on the types, the expected speedups are in the range of x10 - x100.
+
+## Simple usage
+
+Let's take a simple ProtoDef definition and convert it to use the ProtoDef compiler:
+
+ProtoDef:
+```javascript
+const ProtoDef = require('protodef').ProtoDef
+
+// Create a ProtoDef instance
+const proto = new ProtoDef()
+proto.addTypes(require('./protocol.json'))
+
+// Encode and decode a message
+const buffer = proto.createPacketBuffer('mainType', result)
+const result = proto.parsePacketBuffer('mainType', buffer)
+```
+
+ProtoDef Compiler:
+```javascript
+const { ProtoDefCompiler } = require('protodef').Compiler
+
+// Create a ProtoDefCompiler instance
+const compiler = new ProtoDefCompiler()
+compiler.addTypesToCompile(require('./protocol.json'))
+
+// Compile a ProtoDef instance
+const compiledProto = await compiler.compileProtoDef()
+
+// Use it as if it were a normal ProtoDef
+const buffer = compiledProto.createPacketBuffer('mainType', result)
+const result = compiledProto.parsePacketBuffer('mainType', buffer)
+```
+
+## New datatypes
+
+Like the ProtoDef interpreter, the ProtoDef compiler can be extended with custom datatypes. To register a custom type, use the `addTypes(types)` method of the ProtoDef compiler. The `types` parameter is an object with the following structure:
+
+```javascript
+{
+  Read: {
+    'type1': ['native', /* implementation */],
+    'type2': ['context', /* implementation */],
+    'type3': ['parametrizable', /* implementation */],
+    /* ... */
+  },
+
+  Write: {
+    'type1': ['native', /* implementation */],
+    'type2': ['context', /* implementation */],
+    'type3': ['parametrizable', /* implementation */],
+    /* ... */
+  },
+
+  SizeOf: {
+    'type1': ['native', /* implementation */],
+    'type2': ['context', /* implementation */],
+    'type3': ['parametrizable', /* implementation */],
+    /* ... */
+  }
+}
+```
+
+The types can be divided into 3 categories:
+
+### Native Type
+
+A native type is a type read or written by a function that will be called in its original context. Use this when you need access to external definitions.
```

Comment:
a native type is the same as what would happen in the interpreter right ? (might be useful to mention it)

Reply:
yes, in fact you can directly give it the interpreter's functions (like I do for the numeric types). I'll add a note in the doc.

Reply:
done